### PR TITLE
Add ABORT_ON_WASM_EXCEPTIONS #2

### PR DESCRIPTION
**Pull request by tklajnscek**
Copy of https://github.com/emscripten-core/emscripten/pull/11690 as a fresh PR to test if circle-ci tasks will run as discussed here https://github.com/emscripten-core/emscripten/pull/11690#issuecomment-674748154

### CODE REVIEW

File: src/settings.js [hunk 1]
```diff
@@ -1664,6 +1664,19 @@ var WASM2C = 0;
...
+// This makes the Emscripten program behave more like a native program where the OS
+// would terminate the process and no further code can be executed when an unhanled
+// exception (e.g. out-of-bounds memory access) happens.
+// This will instrument all exported functions to catch thrown exceptions and
+// call abort() when they happen. Once the program aborts any exported function calls
+// will fail with a "program has already aborted" exception to prevent calls into
+// code with a potentially corrupted program state.
+// This adds ~250 bytes to the code size in optimized builds and a slight overhead
+// for the extra instrumented function indirection.
+// Disable this if you want to handle the exceptions yourself or save some bytes.
```
Comment (kripken) [inline src/settings.js:1677]: I think this should be off by default. I'd rather not change the behavior for current users (and also increase code size for them).

File: src/preamble.js [hunk 2]
```diff
@@ -802,6 +802,78 @@ function createExportWrapper(name, fixedasm) {
 }
 #endif
 
+#if ABORT_ON_WASM_EXCEPTIONS
+// Instrument all the exported functions to:
+// - abort if an unhandled exception occurs
+// - throw an exception if someone tries to call them after the program has aborted
+// Set settings.ABORT_ON_WASM_EXCEPTIONS for more info.
+__ATINIT__.unshift(function () {
```
Comment (kripken) [inline src/preamble.js:810]: there is a function for this, `addOnInit`.

is this guaranteed to happen before all other constructors etc.? It may be safer to do it the way asyncify does, literally when the wasm exports are received, which is guaranteed to be before any code runs.

File: src/preamble.js [hunk 3]
```diff
@@ -802,6 +802,78 @@ function createExportWrapper(name, fixedasm) {
...
+#if ABORT_ON_WASM_EXCEPTIONS
+// Instrument all the exported functions to:
+// - abort if an unhandled exception occurs
+// - throw an exception if someone tries to call them after the program has aborted
+// Set settings.ABORT_ON_WASM_EXCEPTIONS for more info.
+__ATINIT__.unshift(function () { 
+#if DISABLE_EXCEPTION_CATCHING != 1
+  // With exception catching enabled we only handle exceptions at the top level
+  // so that the exception mechanics can work uninterrupted at the inner level.
+  // Don't do this with exception catching disabled to save some code size.
```
Comment (kripken) [inline src/preamble.js:814]: `DISABLE_EXCEPTION_CATCHING == 2` has a special meaning. is the `!= 1` intentional, to treat 0 the same as 2?

File: src/preamble.js [hunk 4]
```diff
@@ -802,6 +802,78 @@ function createExportWrapper(name, fixedasm) {
...
+            abort("UNH: " + e + ", " + e.stack);
+#endif
+          }
+#if DISABLE_EXCEPTION_CATCHING != 1
+          finally {
+            wrapperDepth -= 1;
+          }
+#endif
+        };
+      }
```
Comment (kripken) [inline src/preamble.js:867]: else on the same line as the `}` please

File: src/preamble.js [hunk 5]
```diff
@@ -802,6 +802,78 @@ function createExportWrapper(name, fixedasm) {
...
+          }
+          
+#if DISABLE_EXCEPTION_CATCHING != 1
+          wrapperDepth += 1;
+#endif
+          try {
+            return original.apply(null, arguments);
+          }
+          catch (e) {
+            if(
```
Comment (kripken) [inline src/preamble.js:842]: space after `if`

File: src/preamble.js [hunk 6]
```diff
@@ -802,6 +802,78 @@ function createExportWrapper(name, fixedasm) {
...
+#endif
+            ) {
+              throw e;
+            }
+            
+#if ASSERTIONS
+            abort("unhandled exception: " + (typeof e === 'object' && e.stack) ? e + ", " + e.stack : e);
+#else
+            abort("UNH: " + e + ", " + e.stack);
+#endif
```
Comment (kripken) [inline src/preamble.js:859]: why is this block needed? I think we can let an unhandled exception remain unhandled, so that the normal browser error event etc. fires.

File: src/preamble.js (hunk 3 above)
Comment (kripken) [reply src/preamble.js:814]: I see, thanks, make ssense.

File: src/preamble.js (hunk 6 above)
Comment (kripken) [reply src/preamble.js:859]: Sorry, I wasn't clear - why do anything when ASSERTIONS are disabled? Is this meant to be used with or without ASSERTIONS?

File: src/preamble.js (hunk 2 above)
Comment (kripken) [reply src/preamble.js]: I'm still worried about this. `Module["asm"]` is an internal detail that we may change from time to time. In particular it's possible that we'll optimize things so that we do not look at that object anymore (we've taken a reference to the export and saved it). Whereas the asyncify way operates on the wasm module exports as they just arrive, which is future-proof.

(Separately, good point that this runs before global ctors, I forgot that these run first.)

Another issue is that export name minification affects `Module["asm"]` too, doesn't it? For example when I build hello world with -O3 I get
```javascript
var _main = Module["_main"] = function() {
 return (_main = Module["_main"] = Module["asm"]["d"]).apply(null, arguments);
};
```
So it's minified as `"d"` in this build.

One option might be to disable export name minification with this option, but that is a little unfortunate.

Another option might be for `callMain` to do something special, like set `wrapperDepth` to negative infinity (so it will always be negative in checks, until a finally block in `callMain` undoes that)?

File: src/preamble.js (hunk 6 above)
Comment (kripken) [reply src/preamble.js:859]: Ah yes, I guess the idea is to handle it separately from ASSERTIONS.

File: src/preamble.js [hunk 7]
```diff
@@ -802,6 +802,78 @@ function createExportWrapper(name, fixedasm) {
...
+              || e === 'longjmp' // rethrow longjmp if enabled
+#endif
+            ) {
+              throw e;
+            }
+            
+#if ASSERTIONS
+            abort("unhandled exception: " + (typeof e === 'object' && e.stack) ? e + ", " + e.stack : e);
+#else
+            abort("UNH: " + e + ", " + e.stack);
```
Comment (kripken) [inline src/preamble.js:858]: we try to be minimalistic without assertions enabled, how about
```suggestion
            abort("UHE");
```

File: src/preamble.js (hunk 2 above)
Comment (tklajnscek) [reply src/preamble.js]: > Another issue is that export name minification affects Module["asm"] too, doesn't it? 

Yeah, that's why I have `var mainFunc = Module["asm"]["main"]` which the minifier modifies to `var mainFunc = Module["asm"]["d"]` as well and it works - the same transform is applied everywhere :)

> Another option might be for callMain to do something special, like set wrapperDepth to negative infinity (so it will always be negative in checks, until a finally block in callMain undoes that)?

Oooh, nice one! Not coming for a JavaScript background I always forget about evil things like this 😎 - this is probably the most bulletproof way of handling this, but it does mean there will be an #ifdefed bit of code in callMain for this there. I'll whip it up and we can review.

File: src/preamble.js (hunk 7 above)
Comment (tklajnscek) [reply src/preamble.js:858]: Yeah I realized that and tried to follow the same pattern - hence the #if. I just thought having actual stack info from where it happened even in shipping builds is nice and worth the ~15 bytes when someone actually has to debug a crash from the wild. The stack trace of the abort calls is not useful, it's the one in `e` that matters.

I'll shorten it to just: abort("UNH "+e.stack) which is smaller, but still super useful.

Comment (tklajnscek) [reply src/preamble.js]: Settled on `abort("UHE:" + [e, e.stack])` since FireFox doesn't print the exception as part of `.stack` like Chromium does.

File: src/preamble.js (hunk 2 above)
Comment (tklajnscek) [reply src/preamble.js]: Changed this now. Let me know if you think there's anything else that needs doing.

File: tests/test_browser.py [hunk 8]
```diff
@@ -2298,7 +2298,7 @@ def test_runtime_misuse(self):
       } catch(e) {
         out('expected fail 1');
         assert(e.toString().indexOf('assert') >= 0); // assertion, not something else
-        ABORT = false; // hackish
+        ABORT = 0; // hackish
```
Comment (kripken) [inline tests/test_browser.py:2301]: are these changes necessary, or aesthetic?

File: src/preamble.js [hunk 9]
```diff
@@ -796,6 +796,75 @@ function createExportWrapper(name, fixedasm) {
...
+              || e === 'longjmp' // rethrow longjmp if enabled
+#endif
+            ) {
+              throw e;
+            }
+            
+#if ASSERTIONS    
+            abort("unhandled exception: " + [e, e.stack]);
+#else
+            abort("UHE:" + [e, e.stack]);
```
Comment (kripken) [inline src/preamble.js:849]: the stack etc. is excessive in a release build, I think, we really care about code size there.
```suggestion
            abort();
```

File: src/preamble.js [hunk 10]
```diff
@@ -796,6 +796,75 @@ function createExportWrapper(name, fixedasm) {
...
+  var instExports = {};
+  for (var name in exports) {
+    (function(name) {
+      var original = exports[name];
+      
+      // Wrap all functions, copy the other symbols.
+      if (typeof original === 'function') {
+        instExports[name] = function() {
+          // Don't allow this function to be called if we're aborted!
+          if (ABORT == 1) {
```
Comment (kripken) [inline src/preamble.js:821]: isn't it also an error if `ABORT == 2`? why not check that as well?

**Comment by kripken** [comment]
I forgot to write, aside from those comments I think this is in good shape, nice work!

File: src/preamble.js (hunk 10 above)
Comment (tklajnscek) [reply src/preamble.js:821]: Before my changes `exit()` would set `abort=true` so any calls after that would fail. 

That's why I changed `var abort` from a bool to a number so that `abort()` sets it to 1 and `exit()` sets it to 2 and we can tell them apart. This was the smallest change I could make.

Here we only want to stop the function call if the program has previously aborted, but not when it exited.

File: tests/test_browser.py (hunk 8 above)
Comment (tklajnscek) [reply tests/test_browser.py:2301]: They're necessary - see my other reply https://github.com/emscripten-core/emscripten/pull/11934#discussion_r481008780

File: src/preamble.js (hunk 9 above)
Comment (tklajnscek) [reply src/preamble.js:849]: I respectfully disagree :)

Without the stack trace it's impossible to find out where the crash happened in a release build because the only thing we'll get back from the field is that there was an abort (and the stack trace from abort()) with the callstack of the actual crash lost.

Getting the stack trace to the actual crash location is essential as there are many bugs we cannot reproduce locally or happen so rarely that it's impractical to do so. 

Since this feature is off-by-default I'm convinced anyone opting into it will gladly take the ~10 byte overhead in build size to get this information which quickly translates into huge real world value (customer satisfaction and qa/dev time).

Comment (kripken) [reply src/preamble.js:849]: I see what you mean now. Makes sense!

In that case, how about removing the difference with assertions? It looks like the familiar pattern of "be super-concise when assertions are off" but isn't, which is what confused me. And it would make the code simpler. As you said, a few bytes in this special mode are not a big deal (but I'm fine with both modes having either the shorter or the longer error).

File: src/preamble.js (hunk 10 above)
Comment (kripken) [reply src/preamble.js:821]: > Before my changes exit() would also set abort=true so any calls to the above after that would fail. But that's not what we want.

Why not? I think that would be correct. After calling `exit()` nothing else should be called, it's an error to do so. Throwing an error is the right behavior.

(In fact we should have assertions that verify that, with ASSERTIONS, so for the exit case the error below is redundant.)

File: src/preamble.js [hunk 11]
```diff
@@ -796,6 +796,75 @@ function createExportWrapper(name, fixedasm) {
...
+          }
+          
+#if DISABLE_EXCEPTION_CATCHING != 1
+          abortWrapperDepth += 1;
+#endif
+          try {
+            return original.apply(null, arguments);
+          }
+          catch (e) {
+            if(
```
Comment (kripken) [inline src/preamble.js:836]: space after `if` please
```suggestion
            if (
```

**Review by kripken — COMMENTED** [review]
This also needs a changelog entry.

File: src/preamble.js (hunk 10 above)
Comment (tklajnscek) [reply src/preamble.js:821]: I'll check, but I think some of the test were failing which is why I made the change. I'll update here.

File: src/preamble.js (hunk 11 above)
Comment (tklajnscek) [reply src/preamble.js:836]: 😱 I did it again!

File: src/preamble.js (hunk 10 above)
Comment (tklajnscek) [reply src/preamble.js:821]: Just re-ran all the tests and I remembered right - anything that uses `EXIT_RUNTIME=1`  and stuff like `printf` fails because there's this in library_syscall.js:

```
#if EXIT_RUNTIME == 1
  fd_write__postset: '__ATEXIT__.push(flush_NO_FILESYSTEM);',
#endif
```

Anything in __ATEXIT__ that runs a WASM function will fail, because `ABORT=true` would be set before running this in `exit()`:
```
ABORT = true;
EXITSTATUS = status;

exitRuntime(); ===> THIS CALLS __ATEXIT__ FUNCTIONS
```
It works fine if I move `ABORT=true` after `exitRuntime`, but I'm worried it will break something else, although doing a quick pass on parts of the code that use `ABORT` it seems safe to do so.
```
ABORT = true;
EXITSTATUS = status;

exitRuntime(); ===> THIS CALLS __ATEXIT__ FUNCTIONS
```
Doing this means I can avoid `ABORT=1/2` and also properly abort when stuff is called after `exit()`.

I'll re-run the tests now and we'll see...

Comment (tklajnscek) [reply src/preamble.js:821]: All tests pass with ABORT=true after exitRuntime() so I'll just go ahead and do that and revert the change of `var ABORT` to a number.

Comment (sbc100) [reply src/preamble.js]: It might be unrelated to this change but programs that abort should probably not call any __ATEXIT__ handlers.   I don't think any user code (including atexit handlers) should run once a program aborts.

File: tests/core/test_abort_on_exception.c [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+EMSCRIPTEN_KEEPALIVE void crash() {
+  printf("crashing\n");
+  throwException();
+}
+
+int main() {
+  return 0;
+} 
+
```
Comment (sbc100) [inline tests/core/test_abort_on_exception.c:25]: Remove trailing newline.

File: src/preamble.js (hunk 10 above)
Comment (tklajnscek) [reply src/preamble.js]: This was actually an `__ATEXIT__` handler from `library_syscall` to flush the streams. So it's library code and it's run _from within_ `exitRuntime()` :)

File: tests/test_core.py [hunk 13]
```diff
@@ -6954,6 +6954,7 @@ def test_embind_unsigned(self):
 
   def test_embind_val(self):
     self.emcc_args += ['--bind']
+    self.set_setting('ABORT_ON_WASM_EXCEPTIONS', 0) # The test handles exceptions itself
```
Comment (sbc100) [inline tests/test_core.py:6957]: Is this still needed given the default is 0?

**Comment by tklajnscek** [comment]
> This also needs a changelog entry.

Done!

File: src/preamble.js (hunk 10 above)
Comment (tklajnscek) [reply src/preamble.js]: Re-read your question and yeah, this was only for regular exit, not when the program aborts. That path doesn't execute any exit handlers.

File: src/postamble.js [hunk 14]
```diff
@@ -189,6 +189,11 @@ function callMain(args) {
     var start = Date.now();
 #endif
 
+#if ABORT_ON_WASM_EXCEPTIONS
+    // See abortWrapperDepth in preamble.js!
+    abortWrapperDepth += 2;
```
Comment (sbc100) [inline src/postamble.js:194]: Dos this no work with `+= 1`?

File: tests/core/test_abort_on_exception.c [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+ * Copyright 2020 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#include <stdio.h>
+#include <emscripten.h>
+
+volatile int* addr = (int*)0xfffffff;
```
[the remark this answers is not in the document]
Comment (tklajnscek) [reply tests/core/test_abort_on_exception.c:11]: Good eyes! Leftover from previous way to cause a crash that doesn't crash in asm.js mode :)

File: tests/test_core.py (hunk 13 above)
Comment (tklajnscek) [reply tests/test_core.py:6957]: Nope, removing.

File: src/postamble.js (hunk 14 above)
Comment (tklajnscek) [reply src/postamble.js:194]: No, because the check in `instrumentWasmExportsWithAbort` is `abortWrapperDepth > 1` - i.e. it will rethrow anything that's not at the top level until it hits top level. So the += 2 basically makes `main` behave as if it's never top level and all the exceptions are rethrown to be handled in `callMain`

File: ChangeLog.md [hunk 16]
```diff
@@ -19,6 +19,13 @@ Current Trunk
 -------------
 - Remove `ALLOC_DYNAMIC` and deprecate `dynamicAlloc`. (#12057, which also
   removes the internal `DYNAMICTOP_PTR` API.)
+- Add `ABORT_ON_EXCEPTIONS` which will abort when an unhandled WASM exception
+  is encountered. This makes the Emscripten program behave more like a native
+  program where the OS would terminate the process and no further code can be
+  executed when an unhanled exception (e.g. out-of-bounds memory access) happens.
```
Comment (kripken) [inline ChangeLog.md:25]:
```suggestion
  executed when an unhandled exception (e.g. out-of-bounds memory access) happens.
```

File: src/settings.js [hunk 17]
```diff
@@ -1606,6 +1606,20 @@ var WASM2C = 0;
 // in a custom location.
 var SEPARATE_DWARF_URL = '';
 
+// Whether the program should abort when an unhandled WASM exception is encountered.
+// This makes the Emscripten program behave more like a native program where the OS
+// would terminate the process and no further code can be executed when an unhanled
```
Comment (kripken) [inline src/settings.js:1611]:
```suggestion
// would terminate the process and no further code can be executed when an unhandled
```

File: src/settings.js [hunk 18]
```diff
@@ -1606,6 +1606,20 @@ var WASM2C = 0;
...
+// exception (e.g. out-of-bounds memory access) happens.
+// This will instrument all exported functions to catch thrown exceptions and
+// call abort() when they happen. Once the program aborts any exported function calls
+// will fail with a "program has already aborted" exception to prevent calls into
+// code with a potentially corrupted program state.
+// This adds ~250 bytes to the code size in optimized builds and a slight overhead
```
Comment (kripken) [inline src/settings.js:1617]: is it 250 even with `--closure 1`? probably less. worth mentioning just an order of magnitude figure in any case, as this may change over time
```suggestion
// This adds a small fixed amount to code size in optimized builds and a slight overhead
```

File: src/settings.js [hunk 19]
```diff
@@ -1606,6 +1606,20 @@ var WASM2C = 0;
...
+// for the extra instrumented function indirection.
+// Enable this if you want Emscripten to handle nuhandled exceptions nicely at the
```
Comment (kripken) [inline src/settings.js:1619]:
```suggestion
// Enable this if you want Emscripten to handle unhandled exceptions nicely at the
```

File: tests/test_core.py [hunk 20]
```diff
@@ -8336,6 +8336,35 @@ def test_emscripten_stack(self):
...
+
+        out("again");
+
+        try {
+          // Try executing some function again
+          _crash();
+        }
+        catch(e) {
+          // Make sure it failed with the expected exception
+          out(e === "program has already aborted!" || e === "DEAD");
```
Comment (kripken) [inline tests/test_core.py:8362]:
```suggestion
          out(e === "program has already aborted!");
```
I think we don't need the DEAD version now?

**Review by kripken — APPROVED** [review]
Ok, I think those are my last comments. Thanks for your patience here!

File: src/settings.js (hunk 17 above)
Comment (tklajnscek) [reply src/settings.js:1611]: I can't believe I how many times I've misspelled this... ಠ_ಠ

**Comment by tklajnscek** [comment]
I believe we're good to go! 🍾

**Comment by kripken** [comment]
Looks good, I fixed a merge conflict and once tests pass will merge. Thanks!
